### ISSUE TITLE
Optimization in Diffusers in comparison with Comfy-UI and Automatic-1111

### ISSUE THREAD

**Issue by tronizx** [huggingface/diffusers#4381]
I'm using diffusers from since months , and found it very good in terms of customization for own deployment services. But few days ago when SDXL 1.0 came out tried it on google colab and the size of model is crashing the session on colab after one or two image generation, while in same specs on google colab Comfy-UI can import two models of SDXL 1.0(Base and Refiner at same time) with still remaining left few ram and gpu memory on google colab session.

So can anyone explains that why there is this much big difference between diffusers and comfy-ui optimizations, that they can import two models of SDXL at same time while a Diffusers crashes the session after importing a single model(Even after applying all other optimization on model as described in documentation of diffusers like, xformers or offloading)?

**Comment by patrickvonplaten** [comment]
Hey @tronizx,

can you copy-paste a code snippet of when `diffusers` crashes for you? `diffusers` is both the fast and most memory efficient framework out there for stable diffusion, but it doesn't apply all the optimizations by default - you have to apply them yourself. 

Maybe you can take a look here: https://huggingface.co/docs/diffusers/optimization/fp16

**Comment by tronizx** [comment]
thank you @patrickvonplaten ,
I'm glad that you replied. About the crash ,When I use SDXL with diffuser it crashes the whole session, with error "Your session is crashed due to high ram usage", and about the link you provided i already been there. here are the optimizations I used while working with diffusers (SDXL) -
-  Half Precision float16
-  enable_model_cpu_offload()
-  Can't use Xformers cause it is already stated in documentation for Pytorch 2.0
-  sequential offload
![Screenshot 2023-08-02 212143](https://github.com/huggingface/diffusers/assets/87223249/8a89e703-085a-472f-85e8-2025ca93f2e2)

This screen shot is captured after generating one image using this code , and it crashed the whole session, while in the same colab and quota in Comfy-UI I can use base model and Refiner model at same time.

**Comment by sayakpaul** [comment]
Could you maybe try out sequential model offload? Also, on Colab, SDPA is already used (an efficient implementation of attention already present in PyTorch 2.0). So no need to use attention slicing.

**Comment by tronizx** [comment]
Thank you so much for reading @sayakpaul ,
as you said i tried it with `pipe.enable_sequential_cpu_offload ()` and got this result-
![Screenshot 2023-08-09 145656](https://github.com/huggingface/diffusers/assets/87223249/be9e94a8-35e9-4bfb-a870-4d39a1dc46d7)

it took 2:11 minutes to generate one image for 50 steps and approx 1:20 minutes with 20 steps. Whereas it way too less in comfy ui look here:- (in ComfyUI)
![Screenshot 2023-08-02 p](https://github.com/huggingface/diffusers/assets/87223249/72445379-2d87-47bf-b534-80617893caaf)

so after getting this results i tried it with torch 2.0 optimization and applied this to pipeline : 
`pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)`
and as you can see it produced error like this, and tried it before with other optimizations and same results:
![Screenshot 2023-08-09 145916](https://github.com/huggingface/diffusers/assets/87223249/5df0b3d4-9b56-4ce4-ab3e-076449ea47d2)
![Screenshot 2023-08-09 145854](https://github.com/huggingface/diffusers/assets/87223249/752ecef1-20ac-4dfa-a04e-51bd845d1416)



I don't why it is producing error in google colab and checked  the version of pytorch too it is equal to 2.0.X

**Comment by sayakpaul** [comment]
You cannot use both at the same time i.e., model offloading and torch compile. Also, a request -- please post full code snippets instead of snapshots so that we can go ahead and try them. 

What happens when you try out xformers, instead of SDPA? 

Cc: @pcuenca for model offloading and compile.

**Comment by tronizx** [comment]
Yeah thanks and  sure i will keep in mind to post full code next time, so is there no way to speed up the inference speed after using sequential offloading?? or may another way to optimize it in low ram but with faster inference?

> You cannot use both at the same time i.e., model offloading and torch compile. Also, a request -- please post full code snippets instead of snapshots so that we can go ahead and try them.
> 
> What happens when you try out xformers, instead of SDPA?
> 
> Cc: @pcuenca for model offloading and compile.
